src/tophu/multiscale.py

```diff
@@ -0,0 +1,465 @@
+    ----------
+    igram : numpy.ndarray
+        The input interferogram. A two-dimensional array.
+    coherence : numpy.ndarray
+        The sample coherence coefficient, with the same shape as the input
+        interferogram.
+    nlooks : float
+        The effective number of looks used to form the input interferogram and
+        coherence.
+    unwrap : UnwrapCallback
+        A callback function used to unwrap the low-resolution interferogram and each
+        high-resolution interferogram tile.
+    downsample_factor : tuple of int
+        The number of looks to take along each axis in order to form the low-resolution
+        interferogram.
+    ntiles : tuple of int
+        The number of tiles along each axis. A pair of integers specifying the shape of
+        the grid of tiles to partition the input interferogram into.
+    tile_overlap : tuple of int or None, optional
+        The overlap between adjacent tiles along each array axis, in samples.
+        (default: None)
+    shape_factor : float, optional
+        The shape factor of the anti-aliasing low-pass filter applied prior to
+        multilooking (the ratio of the width of the combined pass-band and transition
+        band to the pass-band width). A larger shape factor results in a more gradual
+        filter roll-off. (default: 1.5)
+    overhang : float, optional
+        The fraction of the low-pass filter transition bandwidth that extends beyond the
+        Nyquist frequency of the resulting multilooked data. For example, if
+        `overhang=0`, the transition band will be entirely within the Nyquist bandwidth.
+        If `overhang=0.5`, the transition band will centered on the Nyquist frequency.
+        The value must be within the interval [0, 1]. (default: 0.5)
+    ripple : float, optional
+        The maximum allowed ripple amplitude below unity gain in the pass-band of the
+        pre-filter, in decibels. (default: 0.01)
+    attenuation : float, optional
+        The stop-band attenuation of the pre-filter (the difference in amplitude between
+        the ideal gain in the pass-band and the highest gain in the stop-band), in
+        decibels. (default: 40)
+
+    Returns
+    -------
+    unwrapped_phase : numpy.ndarray
+        The unwrapped phase, in radians. An array with the same shape as the input
+        interferogram.
+    conncomp : numpy.ndarray
+        An array of connected component labels, with the same shape as the unwrapped
+        phase.
+    """
+    if igram.shape != coherence.shape:
+        raise ValueError(
+            "shape mismatch: interferogram and coherence arrays must have the same"
+            " shape"
+        )
+    if nlooks < 1.0:
+        raise ValueError("effective number of looks must be >= 1")
+    if any(map(lambda d: d < 1, downsample_factor)):
+        raise ValueError("downsample factor must be >= 1")
+    if any(map(lambda n: n < 1, ntiles)):
+        raise ValueError("number of tiles must be >= 1")
+
```

Comment:
What do you think of the idea of having something like this at this stage:
```python
    if np.prod(ntiles) == 1:
        return unwrap(
            igram=igram,
            corrcoef=coherence,
            nlooks=nlooks,
        )
```

since right now, if you try and do `tiles = (1, 1)`, it'll just filter it and unwrap, then adjust components from filtered to unfiltered version.

(I actually can't get the tile=(1, 1) case to run, since it tells me the `bandwidth` of the filter is > 1 and invalid)

Reply:
Sounds good, I like the idea of special-casing the `ntiles=(1, 1)` case to avoid a bunch of unnecessary steps. I hadn't really considered that use case at all.

>(I actually can't get the tile=(1, 1) case to run, since it tells me the bandwidth of the filter is > 1 and invalid)

Good catch, seems like there's an issue when `downsample_factor` is either `(N, 1)` or `(1, N)`, unless you mess with the filter parameters. In this case, the multilooking & upsampling steps are no-ops along at least one axis, but the low-pass filtering is still performed, which raises an exception with the default filter parameters. Probably, we should skip low-pass filtering for each axis where `downsample_factor` was `1`.

I'll implement the necessary changes and add a few more simple test cases.

Reply:
Oh i just thought of something about this actually-

Maybe the special case should only be if both `ntiles=(1,1)` *and* `downsample_factor=(1,1)` (so neither tiling nor downsampling).

The original version I saw of using downsampled SNAPHU to unwrap higher-res images didn't do tiling, but did run snaphu on extra-multilooked (like in https://www.sciencedirect.com/science/article/pii/S0034425720304338).
That seems like a very useful case to have as an option here.

Reply:
Sounds good to me, I think that makes sense. Thanks for the paper :raised_hands: